### PR TITLE
Adds silicon tipping

### PR DESCRIPTION
**Pull request by BiancaWilkson**
## What Does This PR Do

This PR allows humans to tip over cyborgs, and other silicons after a short delay by clicking on them while in disarm intent. It flips their sprite, crushes people in their way, and immobilizes the cyborg until someone rights them or the borg resists and waits a 20 seconds timer.

It also documents, and makes the SpinAnimation proc slightly more modular. I was originally using that as a base when I changed to using vendor tipping halfway through, with my commits already all mixed up. I'll gladly break atomicity a bit for documenting an old proc.
## Why It's Good For The Game

Thank you to cocacolagua for making the perfect infographic for this.

borg tipping fixed

On a more serious note, having a way to disable cyborgs outside of keeping them permanently flashed or permanently decelled would be a very nice change for both sides. Getting flashed once is already a death sentence for a cyborg anyway, even more so if they have the time to flip you anyway. This at least gives the cyborgs the ability to use their tools, or call for help if the attacker is going to kill them.
## Images of changes

https://github.com/ParadiseSS13/Paradise/assets/42818125/261f5ba2-0be2-42aa-aac9-98d518173215

## Testing

Tipped over a bunch of different silicons, turned them right-side-up or was able to flip myself back over as a borg. The bug in the video seems to be caused by the drone's icon being set back upright when picked up. It's late and I'm tired though.
## Changelog
:cl:
add: Cyborgs can now be tipped over to immobilize them for a short time. Tipped cyborgs can resist to right themselves.
/:cl:

### CODE REVIEW

**Comment by spacegeese** [comment]
best pr of the year! quick, put on that cardborg costume before the silicommunity finds you! oh wait you play binary nvm

on a more serious note what's stopping tots from emagging drones and shoving them on people?

**Comment by Fordoxia** [comment]
Did the tipping do 93% damage to the borg? :/

**Comment by spacegeese** [comment]
> Did the tipping do 93% damage to the borg? :/

that's a big issue for a supposedly nonlethal incapacitating strategy

**Comment by BiancaWilkson** [comment]
> Did the tipping do 93% damage to the borg? :/

I think that's a bug with crashing into someone, I'll get that fixed.

**Comment by Scribble-Sheep** [comment]
Are AI cores able to right themselves?

**Comment by PopeDaveThe3th** [comment]
Additionally, does this have any reactivity to engi cyborgs with their magboots activated?

I will say this overall seems like something people are going to just use to be assholes to Borgs with.

File: code/modules/mob/living/silicon/silicon_defense.dm [hunk 1]
```diff
@@ -61,11 +61,38 @@
...
+	if(do_mob(user, src, tipping_time))
+		var/tip_turf = get_step(src, get_dir(user, src))
+		if(!fall_and_crush(tip_turf, tilt_damage, weaken_time = 0, knockdown_time = 0, angle = HALF_TURN, ignore_gravity = TRUE, rightable = TRUE, untilt_time = tipping_time))
+			is_being_tipped = FALSE
+			return FALSE
+		drop_hat()
+		visible_message(
+			"<span class='notice'>[user] tips [src] over!</span>",
+			"<span class='danger'>[user] tips you over!</span>",
+			"<span class=warning'>You the clatter of metal on the ground.</span>")
```
Comment (Henri215) [inline code/modules/mob/living/silicon/silicon_defense.dm:94]:
```suggestion
			"<span class='warning'>You the clatter of metal on the ground.</span>")
```

Comment (Scribble-Sheep) [reply code/modules/mob/living/silicon/silicon_defense.dm:94]: Is this meant to be "you hear the clatter" or something?

File: code/modules/mob/living/silicon/robot/robot_defense.dm [hunk 2]
```diff
@@ -48,10 +48,29 @@
...
+	..()
+	if(HAS_TRAIT(src, TRAIT_BORG_TIPPED))
+		untilt(src, 20 SECONDS)
+
+/mob/living/silicon/robot/tip_over(mob/living/carbon/human/user)
+	if(HAS_TRAIT(src, TRAIT_BORG_TIPPED))
+		untilt(user, tipping_time)
+		return
+	if(..())
+		ADD_TRAIT(src, TRAIT_BORG_TIPPED, "tipped_over")
```
Comment (GDNgit) [inline code/modules/mob/living/silicon/robot/robot_defense.dm:68]:
```suggestion
		ADD_TRAIT(src, TRAIT_BORG_TIPPED, "tipped_over")
```
trait source this

File: code/modules/mob/living/silicon/robot/drone/maint_drone.dm [hunk 3]
```diff
@@ -23,6 +23,8 @@
 	can_be_hatted = TRUE
 	can_wear_restricted_hats = TRUE
 	/// Cooldown for law syncs
+	tipping_time = 2 SECONDS
```
Comment (GDNgit) [inline code/modules/mob/living/silicon/robot/drone/maint_drone.dm:26]:
```suggestion
	tipping_time = 2 SECONDS
```
would recommend like 5 seconds for this

**Review by GDNgit — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/silicon/robot/robot_defense.dm [hunk 4]
```diff
@@ -48,10 +48,29 @@
...
+		spark_system.start()
+		spawn(0)
+			step_away(src, user, 15)
+			sleep(3)
+			step_away(src, user, 15)
+
+/mob/living/silicon/robot/run_resist()
+	..()
+	if(HAS_TRAIT(src, TRAIT_BORG_TIPPED))
+		untilt(src, 20 SECONDS)
```
Comment (Fordoxia) [inline code/modules/mob/living/silicon/robot/robot_defense.dm:61]: Oh boy, almost half a minute...

File: code/modules/mob/living/silicon/robot/drone/maint_drone.dm [hunk 5]
```diff
@@ -23,6 +23,8 @@
 	can_be_hatted = TRUE
 	can_wear_restricted_hats = TRUE
 	/// Cooldown for law syncs
```
Comment (Oyu07) [inline code/modules/mob/living/silicon/robot/drone/maint_drone.dm:25]: Dont forget to move the comment to its actual var please

File: code/modules/mob/living/silicon/silicon_defense.dm [hunk 6]
```diff
@@ -61,11 +61,38 @@
...
+			tip_over(M)
 		else
 			M.do_attack_animation(src, ATTACK_EFFECT_PUNCH)
 			playsound(loc, 'sound/effects/bang.ogg', 10, 1)
 			visible_message("<span class='notice'>[M] punches [src], but doesn't leave a dent.</span>", \
 						"<span class='notice'>[M] punches [src], but doesn't leave a dent.</span>")
 	return FALSE
+
+/*
+Rotates a silicon's sprite 180 degrees, returns true if successful, false otherwise. All of the un-tipping is handled by the tilted component
```
Comment (Oyu07) [inline code/modules/mob/living/silicon/silicon_defense.dm:76]: Already using multiline
Its getting a tiny bit long, can you make each sentence a new line?

**Review by Oyu07 — COMMENTED** [review]
Comment request
1 wrong comment / important
1 style request / optional

File: code/modules/mob/living/silicon/robot/robot_defense.dm (hunk 4 above)
Comment (DGamerL) [reply code/modules/mob/living/silicon/robot/robot_defense.dm:61]: Please keep reviews for code related complaints, if you want to suggest better options to the Author, you can always copypaste the code into codeblocks

**Comment by Fordoxia** [comment]
> this looks neat for alternate ways of stopping borgs!
> 
> Though the fact they can use tools is a bit worrying to me, what stops them from stun arming the person next to them even after being tilted and then just righting themselves up?

The fact that they need 20 seconds to right themselves, and using tools cancels the timer.

**Comment by BiancaWilkson** [comment]
> > this looks neat for alternate ways of stopping borgs!
> > Though the fact they can use tools is a bit worrying to me, what stops them from stun arming the person next to them even after being tilted and then just righting themselves up?
> 
> The fact that they need 20 seconds to right themselves, and using tools cancels the timer.

Using tools doesn't cancel do_afters, only swapping tools/modules does.

To address the original concern, this was intended to give them some way to not be entirely useless when tipped over.

**Comment by BiancaWilkson** [comment]
> > > this looks neat for alternate ways of stopping borgs!
> > > Though the fact they can use tools is a bit worrying to me, what stops them from stun arming the person next to them even after being tilted and then just righting themselves up?
> > 
> > 
> > The fact that they need 20 seconds to right themselves, and using tools cancels the timer.
> 
> Using tools doesn't cancel do_afters, only swapping tools/modules does.
> 
> To address the original concern, this was intended to give them some way to not be entirely useless when tipped over. Keeping your dis

**Comment by Fordoxia** [comment]
> > > this looks neat for alternate ways of stopping borgs!
> > > Though the fact they can use tools is a bit worrying to me, what stops them from stun arming the person next to them even after being tilted and then just righting themselves up?
> > 
> > 
> > The fact that they need 20 seconds to right themselves, and using tools cancels the timer.
> 
> Using tools doesn't cancel do_afters, only swapping tools/modules does.
> 
> To address the original concern, this was intended to give them some way to not be entirely useless when tipped over.

Why swapping tools, though?

**Comment by BiancaWilkson** [comment]
> > > > this looks neat for alternate ways of stopping borgs!
> > > > Though the fact they can use tools is a bit worrying to me, what stops them from stun arming the person next to them even after being tilted and then just righting themselves up?
> > > 
> > > 
> > > The fact that they need 20 seconds to right themselves, and using tools cancels the timer.
> > 
> > 
> > Using tools doesn't cancel do_afters, only swapping tools/modules does.
> > To address the original concern, this was intended to give them some way to not be entirely useless when tipped over.
> 
> Why swapping tools, though?

It's just how the progress bars work. Sorta like how vampires can strip headsets while batonning someone, but if you try to cuff someone while stripping their headset it gets cancelled when the cuffs leave your hands.

**Comment by Fordoxia** [comment]
> > > > > this looks neat for alternate ways of stopping borgs!
> > > > > Though the fact they can use tools is a bit worrying to me, what stops them from stun arming the person next to them even after being tilted and then just righting themselves up?
> > > > 
> > > > 
> > > > The fact that they need 20 seconds to right themselves, and using tools cancels the timer.
> > > 
> > > 
> > > Using tools doesn't cancel do_afters, only swapping tools/modules does.
> > > To address the original concern, this was intended to give them some way to not be entirely useless when tipped over.
> > 
> > 
> > Why swapping tools, though?
> 
> It's just how the progress bars work. Sorta like how vampires can strip headsets while batonning someone, but if you try to cuff someone while stripping their headset it gets cancelled when the cuffs leave your hands.
It's not, though, there's a setting to disable that. Look at the bolt/shock timers on doors, you can swap hands/tools without interrupting those.

**Comment by Spacemanspark** [comment]
While I concede the value in being able to keep a droid player in the round, my opinion on this is largely the same as it was during the discord conversation we had some time ago on the subject; I feel this feature will be abused to harass droid players more than anything else. Droids/ borgs are already some of the worst treated roles OOCly and ICly, and while that's partially as intended... this could very easily cross that line.

**Review by Sirryan2002 — APPROVED** [review]

**Comment by S34NW** [comment]
Sadly, this PR is 7 days past the deadline and has failed the vote due to 3 objections from design team vs 1 approval.
4 Balance team members approved the PR however design not approving blocks this from progressing.

Some anonymised objection reasons held internally:

- Personally, I don't like this. But I'm willing to see how it goes over in a TM before I outright deny it.
- It's weird and quirky, which isn't a bad thing, but we've already got flashes to stun borgs for a remarkably long time, I don't think we need this, too.
- With how I often see it used, and rarely see tipping and sparing, I'm going to say no at this time.

Thanks for your work on this and feel free to revisit it after discussions with design team!

**Pull request closed without merging**